File: src/services/authService.ts
import axios from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { ENV_CONFIG } from '@/config/env';

interface LoginCredentials {
  email: string;
  password: string;
}

interface RegisterCredentials {
  email: string;
  password: string;
  firstName?: string;
  lastName?: string;
}

interface AuthResponse {
  user: {
    id: string;
    email: string;
    firstName: string;
    lastName: string;
    avatar?: string;
    role: string;
  };
  accessToken: string;
  refreshToken: string;
}

interface SocialAuthResponse {
  user: {
    id: string;
    email: string;
    name: string;
    avatar?: string;
    provider: 'google' | 'facebook' | 'apple';
  };
  accessToken: string;
  refreshToken: string;
}

class AuthService {
  private baseURL: string;

  constructor() {
    this.baseURL = ENV_CONFIG.API_BASE_URL;
  }

  // Email/Password Authentication
  async login(credentials: LoginCredentials): Promise<AuthResponse> {
    try {
      const response = await axios.post(`${this.baseURL}/auth/login`, credentials);
      
      // Store tokens
      await this.storeTokens(response.data.accessToken, response.data.refreshToken);
      
      return response.data;
    } catch (error) {
      console.error('Login error:', error);
      throw new Error('Login failed. Please check your credentials.');
    }
  }

  async register(credentials: RegisterCredentials): Promise<AuthResponse> {
    try {
      const response = await axios.post(`${this.baseURL}/auth/register`, credentials);
      
      // Store tokens
      await this.storeTokens(response.data.accessToken, response.data.refreshToken);
      
      return response.data;
    } catch (error) {
      console.error('Registration error:', error);
      throw new Error('Registration failed. Please try again.');
    }
  }

  // Social Authentication
  async googleLogin(googleToken: string): Promise<SocialAuthResponse> {
    try {
      const response = await axios.post(`${this.baseURL}/auth/google`, {
        token: googleToken,
      });
      
      await this.storeTokens(response.data.accessToken, response.data.refreshToken);
      
      return response.data;
    } catch (error) {
      console.error('Google login error:', error);
      throw new Error('Google login failed. Please try again.');
    }
  }

  async facebookLogin(facebookToken: string): Promise<SocialAuthResponse> {
    try {
      const response = await axios.post(`${this.baseURL}/auth/facebook`, {
        token: facebookToken,
      });
      
      await this.storeTokens(response.data.accessToken, response.data.refreshToken);
      
      return response.data;
    } catch (error) {
      console.error('Facebook login error:', error);
      throw new Error('Facebook login failed. Please try again.');
    }
  }

  async appleLogin(appleToken: string): Promise<SocialAuthResponse> {
    try {
      const response = await axios.post(`${this.baseURL}/auth/apple`, {
        token: appleToken,
      });
      
      await this.storeTokens(response.data.accessToken, response.data.refreshToken);
      
      return response.data;
    } catch (error) {
      console.error('Apple login error:', error);
      throw new Error('Apple login failed. Please try again.');
    }
  }

  // Token Management
  async storeTokens(accessToken: string, refreshToken: string): Promise<void> {
    try {
      await AsyncStorage.multiSet([
        ['accessToken', accessToken],
        ['refreshToken', refreshToken],
      ]);
    } catch (error) {
      console.error('Error storing tokens:', error);
    }
  }

  async getAccessToken(): Promise<string | null> {
    try {
      return await AsyncStorage.getItem('accessToken');
    } catch (error) {
      console.error('Error getting access token:', error);
      return null;
    }
  }

  async getRefreshToken(): Promise<string | null> {
    try {
      return await AsyncStorage.getItem('refreshToken');
    } catch (error) {
      console.error('Error getting refresh token:', error);
      return null;
    }
  }

  async refreshAccessToken(): Promise<string | null> {
    try {
      const refreshToken = await this.getRefreshToken();
      if (!refreshToken) {
        throw new Error('No refresh token available');
      }

      const response = await axios.post(`${this.baseURL}/auth/refresh`, {
        refreshToken,
      });

      const newAccessToken = response.data.accessToken;
      await AsyncStorage.setItem('accessToken', newAccessToken);

      return newAccessToken;
    } catch (error) {
      console.error('Token refresh error:', error);
      await this.logout();
      return null;
    }
  }

  async logout(): Promise<void> {
    try {
      await AsyncStorage.multiRemove(['accessToken', 'refreshToken', 'user']);
    } catch (error) {
      console.error('Logout error:', error);
    }
  }

  async isAuthenticated(): Promise<boolean> {
    try {
      const accessToken = await this.getAccessToken();
      return !!accessToken;
    } catch (error) {
      console.error('Authentication check error:', error);
      return false;
    }
  }

  // Password Reset
  async requestPasswordReset(email: string): Promise<void> {
    try {
      await axios.post(`${this.baseURL}/auth/forgot-password`, { email });
    } catch (error) {
      console.error('Password reset request error:', error);
      throw new Error('Failed to send password reset email.');
    }
  }

  async resetPassword(token: string, newPassword: string): Promise<void> {
    try {
      await axios.post(`${this.baseURL}/auth/reset-password`, {
        token,
        password: newPassword,
      });
    } catch (error) {
      console.error('Password reset error:', error);
      throw new Error('Failed to reset password.');
    }
  }

  // Email Verification
  async verifyEmail(token: string): Promise<void> {
    try {
      await axios.post(`${this.baseURL}/auth/verify-email`, { token });
    } catch (error) {
      console.error('Email verification error:', error);
      throw new Error('Failed to verify email.');
    }
  }

  async resendVerificationEmail(email: string): Promise<void> {
    try {
      await axios.post(`${this.baseURL}/auth/resend-verification`, { email });
    } catch (error) {
      console.error('Resend verification error:', error);
      throw new Error('Failed to resend verification email.');
    }
  }
}

export default new AuthService();

File: lib/data/services/auth_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';
import 'package:logger/logger.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../../core/config/config.dart';
import '../../core/constants/storage_keys.dart';
import '../models/user.dart';

class AuthService {
  AuthService({
    required Dio dio,
    required SharedPreferences prefs,
    required FlutterSecureStorage secureStorage,
    required Logger logger,
  }) : _dio = dio,
       _prefs = prefs,
       _secureStorage = secureStorage,
       _logger = logger {
    _dio.options.baseUrl = Config.apiBaseUrl;
    _dio.options.connectTimeout = Config.apiTimeout;
    _dio.options.receiveTimeout = Config.apiTimeout;
  }
  final Dio _dio;
  final SharedPreferences _prefs;
  final FlutterSecureStorage _secureStorage;
  final Logger _logger;

  // Email/Password Authentication
  Future<AuthResponse> login({
    required String identifier,
    required String password,
  }) async {
    try {
      // Send as 'email' field for backend compatibility
      // The backend should accept both email and username in this field
      final response = await _dio.post<Map<String, dynamic>>(
        '/auth/login',
        data: {'email': identifier, 'password': password},
      );

      _logger.d('[AUTH] Login response received');

      // Use fromBackend to handle NestJS response format
      final authResponse = AuthResponse.fromBackend(response.data!);

      if (authResponse.success && authResponse.tokens != null) {
        await _storeTokens(authResponse.tokens!);
      }

      return authResponse;
    } on DioException catch (e) {
      return AuthResponse(
        success: false,
        error: _extractErrorMessage(e) ??
            'Login failed. Please check your credentials.',
      );
    } on Exception {
      return const AuthResponse(
        success: false,
        error: 'An unexpected error occurred.',
      );
    }
  }

  Future<AuthResponse> register({
    required String email,
    required String password,
    String? firstName,
    String? lastName,
  }) async {
    try {
      final response = await _dio.post<Map<String, dynamic>>(
        '/auth/register',
        data: {
          'email': email,
          'password': password,
          if (firstName != null) 'firstName': firstName,
          if (lastName != null) 'lastName': lastName,
        },
      );

      // Use fromBackend to handle NestJS response format
      final authResponse = AuthResponse.fromBackend(response.data!);

      if (authResponse.success && authResponse.tokens != null) {
        await _storeTokens(authResponse.tokens!);
      }

      return authResponse;
    } on DioException catch (e) {
      return AuthResponse(
        success: false,
        error: _extractErrorMessage(e) ??
            'Registration failed. Please try again.',
      );
    } on Exception {
      return const AuthResponse(
        success: false,
        error: 'An unexpected error occurred.',
      );
    }
  }

  // Social Authentication
  Future<SocialAuthResult> googleLogin(String googleToken) async {
    try {
      final response = await _dio.post<Map<String, dynamic>>(
        '/auth/google',
        data: {'token': googleToken},
      );

      // Use fromBackend to handle NestJS response format
      final authResult = SocialAuthResult.fromBackend(response.data!);

      if (authResult.success && authResult.tokens != null) {
        await _storeTokens(authResult.tokens!);
      }

      return authResult;
    } on DioException catch (e) {
      return SocialAuthResult(
        success: false,
        error:
            (e.response?.data['message'] as String?) ??
            'Google login failed. Please try again.',
      );
    } on Exception {
      return const SocialAuthResult(
        success: false,
        error: 'An unexpected error occurred.',
      );
    }
  }

  Future<SocialAuthResult> facebookLogin(String facebookToken) async {
    try {
      final response = await _dio.post<Map<String, dynamic>>(
        '/auth/facebook',
        data: {'token': facebookToken},
      );

      // Use fromBackend to handle NestJS response format
      final authResult = SocialAuthResult.fromBackend(response.data!);

      if (authResult.success && authResult.tokens != null) {
        await _storeTokens(authResult.tokens!);
      }

      return authResult;
    } on DioException catch (e) {
      return SocialAuthResult(
        success: false,
        error:
            (e.response?.data['message'] as String?) ??
            'Facebook login failed. Please try again.',
      );
    } on Exception {
      return const SocialAuthResult(
        success: false,
        error: 'An unexpected error occurred.',
      );
    }
  }

  Future<SocialAuthResult> appleLogin(String appleToken) async {
    try {
      final response = await _dio.post<Map<String, dynamic>>(
        '/auth/apple',
        data: {'token': appleToken},
      );

      // Use fromBackend to handle NestJS response format
      final authResult = SocialAuthResult.fromBackend(response.data!);

      if (authResult.success && authResult.tokens != null) {
        await _storeTokens(authResult.tokens!);
      }

      return authResult;
    } on DioException catch (e) {
      return SocialAuthResult(
        success: false,
        error:
            (e.response?.data['message'] as String?) ??
            'Apple login failed. Please try again.',
      );
    } on Exception {
      return const SocialAuthResult(
        success: false,
        error: 'An unexpected error occurred.',
      );
    }
  }

  /// Extract error message from backend response (supports 'message' and 'error' fields)
  static String? _extractErrorMessage(DioException e) {
    final data = e.response?.data;
    if (data is Map<String, dynamic>) {
      final message = data['message'];
      if (message is String) {
        return _cleanMessage(message);
      }
      if (message is List) {
        return _cleanMessage(message.join(', '));
      }
      final error = data['error'];
      if (error is String) {
        return _cleanMessage(error);
      }
    }
    return null;
  }

  /// Remove technical prefixes from backend error messages
  static String _cleanMessage(String msg) => msg
        .replaceFirst(RegExp(r'^Validation failed:\s*'), '')
        .replaceFirst(RegExp(r'^Error:\s*'), '');

  // Token Management
  Future<void> _storeTokens(AuthTokens tokens) async {
    await _secureStorage.write(
      key: StorageKeys.accessToken,
      value: tokens.accessToken,
    );
    await _secureStorage.write(
      key: StorageKeys.refreshToken,
      value: tokens.refreshToken,
    );
  }

  Future<String?> getAccessToken() async =>
      _secureStorage.read(key: StorageKeys.accessToken);

  Future<String?> getRefreshToken() async =>
      _secureStorage.read(key: StorageKeys.refreshToken);

  Future<String?> refreshAccessToken() async {
    try {
      final refreshToken = await getRefreshToken();
      if (refreshToken == null) {
        throw Exception('No refresh token available');
      }

      final response = await _dio.post<Map<String, dynamic>>(
        '/auth/refresh',
        data: {'refreshToken': refreshToken},
      );

      final newAccessToken = response.data?['accessToken'] as String?;
      if (newAccessToken == null) {
        throw Exception('No access token in refresh response');
      }
      await _secureStorage.write(
        key: StorageKeys.accessToken,
        value: newAccessToken,
      );

      return newAccessToken;
    } on Exception {
      await logout();
      return null;
    }
  }

  Future<void> logout() async {
    await _secureStorage.delete(key: StorageKeys.accessToken);
    await _secureStorage.delete(key: StorageKeys.refreshToken);
    await _prefs.remove(StorageKeys.user);
  }

  Future<bool> isAuthenticated() async {
    final accessToken = await getAccessToken();
    return accessToken != null && accessToken.isNotEmpty;
  }

  // Password Reset
  Future<bool> requestPasswordReset(String email) async {
    try {
      await _dio.post<void>('/auth/forgot-password', data: {'email': email});
      return true;
    } on Exception {
      return false;
    }
  }

  Future<bool> resetPassword({
    required String token,
    required String newPassword,
  }) async {
    try {
      await _dio.post<void>(
        '/auth/reset-password',
        data: {'token': token, 'password': newPassword},
      );
      return true;
    } on Exception {
      return false;
    }
  }

  // Email Verification
  Future<bool> verifyEmail(String token) async {
    try {
      await _dio.post<void>('/auth/verify-email', data: {'token': token});
      return true;
    } on Exception {
      return false;
    }
  }

  Future<bool> resendVerificationEmail(String email) async {
    try {
      await _dio.post<void>(
        '/auth/resend-verification',
        data: {'email': email},
      );
      return true;
    } on Exception {
      return false;
    }
  }
}
